### PR TITLE
Sign in with Apple

### LINKED ISSUE #1117: Support apple id login

### PR DESCRIPTION
**Pull request by louischan-oursky**
fix #1117 

The other platforms integration is just OIDC.

The iOS integration deviates from the authorization code flow so I added two endpoints.
In particular, we pass both the nonce and the hashed nonce to the client while in the authorization code flow, we pass the nonce via HTTP only cookie.

The corresponding iOS integration is here. https://gist.github.com/louischan-oursky/6ca2f70eee957343be7ff3d2eb7d4107

### CODE REVIEW

File: pkg/auth/dependency/sso/azureadv2.go [hunk 1]
```diff
@@ -135,12 +135,12 @@ func (f *Azureadv2Impl) OpenIDConnectGetAuthInfo(r OAuthAuthorizationResponse, s
...
 	c, err := f.getOpenIDConfiguration()
 	if err != nil {
-		err = errors.Newf("failed to get OIDC discovery document: %w", err)
+		err = NewSSOFailed(NetworkFailed, "failed to get OIDC discovery document")
 		return
 	}
 	keySet, err := f.getKeys(c.JWKSUri)
 	if err != nil {
-		err = errors.Newf("failed to get OIDC JWKS: %w", err)
+		err = NewSSOFailed(NetworkFailed, "failed to get OIDC JWTs")
```
Comment (kiootic) [inline pkg/auth/dependency/sso/azureadv2.go]: It should be JWKs

File: pkg/auth/dependency/sso/apple.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+		err = NewSSOFailed(InvalidParams, "no nonce")
+		return
+	}
+	if subtle.ConstantTimeCompare([]byte(hashedNonce), []byte(crypto.SHA256String(r.Nonce))) != 1 {
+		err = NewSSOFailed(SSOUnauthorized, "invalid nonce")
+		return
+	}
+
+	// Verify the issuer
+	if !mapClaims.VerifyIssuer("https://appleid.apple.com", true) {
```
Comment (kiootic) [inline pkg/auth/dependency/sso/apple.go]: The referenced document says the issuer should 'contains' the constant string. I'm not sure if it'll be some other strings like `https://appleid.apple.com/abc` in future.

File: pkg/auth/handler/sso/apple_request.go [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+		p.MergeRealm = password.DefaultRealm
+	}
+
+	if p.OnUserDuplicate == "" {
+		p.OnUserDuplicate = model.OnUserDuplicateDefault
+	}
+}
+
+func (p *AppleRequestPayload) Validate() []validation.ErrorCause {
+	if !p.PasswordAuthProvider.IsRealmValid(p.MergeRealm) {
```
Comment (kiootic) [inline pkg/auth/handler/sso/apple_request.go]: We should validate the provider is Apple ID.

File: pkg/auth/handler/sso/apple_response.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+	err = hook.WithTx(h.HookProvider, h.TxContext, func() (err error) {
+		result, err = h.Handle(payload)
+		return
+	})
+	h.AuthnSessionProvider.WriteResponse(w, result, err)
+}
+
+func (h *AppleResponseHandler) Handle(payload *AppleResponseRequest) (result interface{}, err error) {
+	if h.OAuthProvider == nil {
```
Comment (kiootic) [inline pkg/auth/handler/sso/apple_response.go]: We should validate the provider is Apple ID.

File: pkg/auth/dependency/sso/apple.go (hunk 2 above)
Comment (louischan-oursky) [reply pkg/auth/dependency/sso/apple.go]: The wording is a bit confusing. In my real testing, the value of issuer is exactly `https://appleid.apple.com`. If I change to use `strings.HasPrefix`, do you have any concerns?

Comment (kiootic) [reply pkg/auth/dependency/sso/apple.go]: My concern is that the document explicitly used the word 'contains' for `iss` field, and the word 'is' is used for exact matching the `aud` field. It seems to me that they anticipated the need to change it in future. I think using `strings.Contains` would be safer.

**Comment by louischan-oursky** [comment]
@kiootic Removed the two endpoints and added UXModeManual.

**Pull request merged**